### ISSUE TITLE
Generational GC

### ISSUE THREAD

**Issue by DemiMarie** [Marwes/gluon#175]
Gluon may want to get a generational moving GC, for performance.

**Comment by Marwes** [comment]
Since this is an optimization we definitely need to add more benchmarks for different workloads as a first step (its going to be an improvement but measuring how much is also improvement).

The current collector is not a straight a straight mark and sweep collector however as to some extent it is possible to limit how much is scanned in each collection by spawning new threads which may give similiar improvements to a generational collector https://github.com/Marwes/gluon/blob/e9f9de9734428da3e8afcf15d027146f1fb58174/vm/src/gc.rs#L110-L129.

**Comment by brendanzab** [comment]
Would be interesting to look into how best to do a GC, seeing as Gluon is meant to be pretty lightweight. We might want some way to control when collection happens, and prevent pauses where in critical places. Might want to look at how luajit does it. But bear in mind that we have the advantage of purity, which could make some things much simpler. (see GHC's collector)

**Comment by DemiMarie** [comment]
@brendanzab I did not know that Gluon is purely functional!  Does it provide anything analogous to the `IO` or `ST` monad, or is it more like Erlang?  I feel like mutable arrays – at least of primitive types – would be helpful for some tasks, such as building strings using more efficient data structures than linked lists.  Haskell might be a good source of inspiration for libraries.

If everything is immutable, then concurrent compacting collection is easy.

**Comment by brendanzab** [comment]
Actually, I think it has references. We do have [ST](https://github.com/Marwes/gluon/blob/master/std/state.glu) and [IO](https://github.com/Marwes/gluon/blob/master/std/prelude.glu#L509-L512). I would like us to get a nice effect system at some stage though, which will mean we can avoid the insanity of monad transformers and free monads - ugh.

**Comment by brendanzab** [comment]
Feel free to chat with us on https://gitter.im/Marwes/gluon by the way!

**Comment by Marwes** [comment]
For the record, a copying collector will cause some headaches when it comes to FFI (which will be common) since any pointers to gluon values passed to a native function must be in a stable location for at least the duration of the call. Been looking for a satisfying solution for this but so far I have only found problems.

**Comment by DemiMarie** [comment]
One solution is to only pass indirect handles (pointers to stack/heap
allocated variables which can't move).  This might require some ugly
callback-based APIs OR a heap-allocated handle table.  It may also be
slower.

**Comment by Marwes** [comment]
Yeah, forcing the use of handles is far from ideal as it prevents passing `&str`, `&[u8]`, etc directly to native functions which is really convenient. 

To be fair, one could 'lock' a handle for a scope where it guarantees that no garbage collection may be run, and allow references to the value in the handle during that scope (such as with `RefCell`). But then we might as well not force the use of handles and just track whether a reference to some moveable data exists.

**Comment by DemiMarie** [comment]
What about generational quad-color mark-region GC?  LuaJIT 3.0 will use it;
the algorithm can be found somewhere on https://luajit.org.

**Comment by Marwes** [comment]
Actually meant to add a link to it but I forgot. It has no implementation at all currently so its hard to tell how good it actually is, definitely interesting though. http://wiki.luajit.org/New-Garbage-Collector
